Allow configuration of how builds are sorted in web UI

AdamWill opened the pull request:
Currently, openQA assumes that `BUILD` values can be sorted in
'version number' style, when sorting the lists of builds for the
web UI index page and job group overview pages. SUSE follows
this convention, but Fedora doesn't and never has. Our compose
`BUILD` values are productmd 'compose IDs', like this:

Fedora-26-20170301.n.0
Fedora-Rawhide-20170301.n.0

and we have recently started testing updates, whose `BUILD`
values look like this:

FEDORA-2017-561942797a

Where the final string is just a randomly-generated identifier,
it cannot be relied upon to sort in any way at all. It's not
really plausible to change these `BUILD` values into ones that
can be sorted like version numbers, so instead, let's make it
possible to specify at the job group level whether `BUILD`
values for jobs in that group can be sorted as versions or not.
If not, `compute_build_results` will stick with the initial
sort order, which is to use the time a job was most recently
created for each build as the key.

Note: I haven't tested this yet, not sure I want to do a test schema version bump on our staging instance and my pet instance isn't working right now. Will try and test it one way or another soon, but wanted to send the PR for general comments on the idea before I go any further (and to see if the tests pass).

AdamWill commented:
Also note, I'm not sure why the schema upgrade decided to change something about the `jobs_fk_assigned_worker_id` constraint when I didn't touch that at all, but I don't know if I want to tweak it manually.

AdamWill commented:
@Martchus thoughts?

AdamWill commented:
So I tested this a bit now and the feature works, but the admin UI config doesn't. I think it's because when using a checkbox like this, `serialize()` doesn't store the value in any way that gets sent back in the `PUT` request, but my javascript is awful so I really don't know how to fix this...lil help anyone? I note @coolo is doing some checkbox stuff in `templates/admin/user/index.html.ep` - can you help with this? Thanks!

AdamWill commented:
Well, found a way to make it work, I think. I don't think using a checkbox can work as no value will be `serialize`d when the checkbox is unchecked, so radio buttons it is, unless anyone knows a better way.

Martchus commented on templates/admin/group/group_property_editor.html.ep:
```diff
@@ -23,6 +23,17 @@
+                    <div class="col-sm-10">
+                        <input type="radio" id="editor-build-version-sort" name="build_version_sort" value="1"
+                            %= 'checked' if $group->build_version_sort
+                            > Yes
```
The yes/no text should be the label for the radio button, not the text at the left. But the use of a radio button should be ok in general.

Martchus commented on templates/admin/group/group_property_editor.html.ep:
```diff
@@ -23,6 +23,17 @@
+                        <input type="radio" id="editor-build-version-sort" name="build_version_sort" value="1"
+                            %= 'checked' if $group->build_version_sort
+                            > Yes
+                        <input type="radio" id="editor-build-version-sort" name="build_version_sort" value="0"
```
You can't use the same ID twice in a document.

AdamWill commented:
OK, how's that? The buttons still don't look super awesome (they're kinda close together and I dunno why the label / button alignment is a bit off...), but I think it's what you wanted?

AdamWill commented:
BTW, on the schema upgrade: I hand-edited the deploy and upgrade files not to change the `jobs_fk_assigned_worker_id` constraint at all, since I couldn't see any reason why they should. My best guess is it happened because I'm using different versions of the relevant perl libraries (I generated the files on Fedora Rawhide, so they're pretty new versions). Let me know if that was the wrong choice.

Martchus commented:
The change regarding the label was about behaviour, not the appearance. I can check when I'm at work again how it looks like, but I'm sure with a little bit css it should look acceptable.

Newer version can be the reason for the db upgrade. Or we just missed it last time the schema was modified and missing it has no impact. I can check that when I'm in the office again.

AdamWill commented:
It looks like this. If you can pretty it up a bit, go for it - I suck at web design. :P

For the schema thing, I looked through all the changes since the previous schema version bump and couldn't see any that do anything to that constraint, so I'm guessing it's the different versions of the DBIx libs. I suppose you could just try bumping the schema version in current master, running the deploy/upgrade script generation commands, and see if you get the same thing...

![screenshot from 2017-03-02 12-30-27](https://cloud.githubusercontent.com/assets/916551/23525655/0f6d7cc6-ff44-11e6-8ae1-85031c9232ac.png)

AdamWill commented:
ping?

okurz commented:
I don't quite get the question "Sort BUILDs as versions?". The alternative is not clear. Also the other points are described in imperative mood -> delete '?'
Maybe also add the help icon with a slightly longer description.

AdamWill commented:
Yeah, I was having trouble trying to phrase it concisely. Note there is a tooltip which explains in a bit more detail (as for all the others). I can have a go at trying to reword it, though.

okurz commented:
@AdamWill well, you are the native, you will manage ;-)

AdamWill commented:
other than that, does it look good to you? I'm hoping to get it signed off / merged soon, because I can't safely deploy this downstream without upstream approval because of the schema bump - if I do this schema bump downstream, but then the schema bump upstream winds up being different, we're screwed :)

okurz commented on t/22-dashboard.t:
```diff
@@ -410,7 +417,14 @@ subtest 'job groups with multiple version and builds' => sub {
+    $get         = $t->get_ok('/group_overview/' . $group->id)->status_is(200);
+    @h4          = $get->tx->res->dom->find("div.no-children h4 a")->map('text')->each;
+    @build_names = reverse @build_names;
+    is_deeply(\@h4, \@build_names, 'builds shown sorted by job creation time') || diag explain @h4;
```
please extract helper method

Martchus commented:
I haven't been in the office yet to actually test it but the code looks good now. Labels are use correctly and there's a test. Note that only @coolo and @aaannz can merge PR.

okurz commented on t/deploy.t:
```diff
@@ -81,10 +81,6 @@ ok($dh->version_storage->database_version, 'DB deployed');
-# check systemuser exists after upgrade
-my $user = $schema->resultset('Users')->find({username => 'system'});
-ok($user->id, 'system user does exists');
-
```
why delete?

aaannz replied:
TLDR; I'm fine with removing it

This test is now useless. I added it to catch the system user creation during migration (where indeed was a bug), but since then creating system user moved to deployment subroutine, this means even the first `$version - 1` schema deployment in this test creates a system user.

The presence of system user on deployed system is checked in [t/06-users.t](https://github.com/os-autoinst/openQA/blob/master/t/06-users.t#L40)

aaannz replied:
Question is if this change is related to this PR. That I don't think so.

AdamWill replied:
It's actually explained in the current version of the commit message. It *is* related to the commit, though it's not obvious at first - it made me scratch my head for a while till I figured it out.

The schema upgrade test works by deploying the schema at version "current version minus 2", then upgrading it. This particular assertion assumes that the 'system' user creation will happen as part of that upgrade process. However, the creation of the 'system' user *as part of schema upgrade* only occurs when upgrading from schema version 50 or earlier.

So this breaks purely because this commit *bumps the schema version*. As long as the schema version was 52, the schema upgrade test would deploy schema version 50 and upgrade to 52, and that process *would* include the 'system' user creation, and the assertion would pass. However, since this commit bumps the schema version to 53, the upgrade test now deploys schema version 51 and upgrades to version 53. The 'system' user is *not* created as part of that transaction, and the assertion fails.

So I think it does belong in this commit.

okurz replied:
very convincing :-)

okurz commented:
> Question is if this change is related to this PR. That I don't think so.

I would not mind if it stays within this PR but @AdamWill please create 
another commit for this.

okurz commented:
so test change is fine.

LGTM

AdamWill commented:
That last push was just a rebase on current master.

okurz commented:
LGTM
